### PR TITLE
fix(BV): Internalize sign_extend and repeat

### PR DESCRIPTION
**Pull request by bclement-ocp**
Currently, we convert `sign_extend` and `repeat` to the appropriate concatenations and extractions at the term level. This has the advantage that it is simple, but the inconvenient that we create a potentially large amount of terms that do not appear in the original problem (proportional to the size of the sign extension or repetition).

In addition to preventing obvious simplifications with extractions, the creation of many terms clutters the union-find which drastically slows down the solver in presence of sign extensions to large bit-widths.

Note: this impacts some of the Décysif datasets.

### CODE REVIEW

**Comment by bclement-ocp** [comment]
> Note: this impacts some of the Décysif datasets.

More precisely, this patch is on average 6x faster and solves 10% more problems than current `next` on the J3 dataset.

File: tests/bitv/testfile-sign-extend-003.smt2 [hunk 1]
```diff
@@ -0,0 +1,4 @@
+(set-logic ALL)
+(declare-const x (_ BitVec 4))
+(assert (distinct ((_ extract 3 2) ((_ sign_extend 4) x)) ((_ extract 3 2) x)))
+(check-sat)
```
Comment (Halbaroth) [inline tests/bitv/testfile-sign-extend-003.smt2]: This test is identical with `testfile-sign-extend-002.smt2`.

File: src/lib/structures/expr.ml [hunk 2]
```diff
@@ -3097,17 +3097,17 @@ module BV = struct
       (Sy.Op (Sy.Extract (j, i))) [s] (Ty.Tbitv (i - j + 1))
 
   (* Other operations *)
-  let rec repeat i t =
+  let repeat i t =
     assert (i >= 1);
```
Comment (Halbaroth) [inline src/lib/structures/expr.ml]: We may use `invalid_arg` here.

**Review by Halbaroth — APPROVED** [review]
LGTM 
Feel free to apply my suggestions.

File: tests/bitv/testfile-sign-extend-003.smt2 (hunk 1 above)
Comment (bclement-ocp) [reply tests/bitv/testfile-sign-extend-003.smt2]: Oops, this one was intended to test extraction of the high bits. Fixed.

File: src/lib/structures/expr.ml (hunk 2 above)
Comment (bclement-ocp) [reply src/lib/structures/expr.ml]: Done

**Pull request merged**